### ISSUE TITLE
Connection pool design advice

### ISSUE THREAD

**Issue by adtrevor** [swift-server/async-http-client#25]
I’m seeking for design advice on how to build a connection pool, I would like to spend some time implementing it and hopefully contribute them if I get good results.
I don't know exactly where to start so I thought it might be good to expose what I understand on how the connection pool should work, and show design ideas I found by exploring already existing implementations.

First, the general concept of the connection pool: on each new request, the client asks its connection pool for a connection to the server. When it receives a new connection request, the pool has essentially three choices:
- Use an already opened, unused connection
- Open a new connection if no other connection is available
- If there are already too many concurrent connections to the server, wait until an already opened connection becomes available
In practice the connection is a NIO `Channel` that is either directly a TCP connection in the case of HTTP/1.1, or a stream for an HTTP/2 multiplexed connection.

After some research, I stumbled upon the Apache Foundation `HttpClient`. It uses an [`HttpClientConnectionManager` interface](https://hc.apache.org/httpcomponents-client-ga/httpclient/apidocs/org/apache/http/conn/HttpClientConnectionManager.html) that lets the user provide the manager they wish. A `PoolingHttpClientConnectionManager` class implements this interface.
I wonder wether the principle of a user provided connection manager makes sense for the Swift NIO HTTP client? In the case of the Apache, I see the other class that implements `HttpClientConnectionManager` is `BasicHttpClientConnectionManager`, a connection manager that maintains only a single active connection at a time, but if I understand it correctly, this manager purpose is to enable to have one client per thread, which doesn’t seem appliable/useful with NIO? We can also imagine concurrent yet unpooled connection management, which is how the client currently works, but I don’t see the advantage of this over a pooled connection management?

Speaking more specifically about the connection pool, I have a few questions:
- How to give back a connection to the pool? Should we pass it as an argument to a pool method (such as [this](https://hc.apache.org/httpcomponents-client-ga/httpclient/apidocs/org/apache/http/conn/HttpClientConnectionManager.html#releaseConnection(org.apache.http.HttpClientConnection,%20java.lang.Object,%20long,%20java.util.concurrent.TimeUnit))) , or could we have a design where the pool passes an object (maybe a promise?) to the client on which it calls the appropriate method when it has finished leasing the connection?
- How should we implement the thread safety of the pool? The already existing APIs based on Future/Promises should make this easy for many parts, but the data structure storing the connections must still be protected. Regarding the threading model of NIO, is this a good practice to hold a reference to a specific event loop and to use .execute to ensure what we want is executed on the right thread?
- Speaking of the data structure storing the connections, could this be as simple as a dictionary that keys the connections per host/scheme?
- I saw that some pools have a global cleanup delay where the manager will loop over all opened connections to see which ones have been idle for too long in order to evict them. How does this « global cleanup interval » strategy compare to scheduling timeouts per connection using NIO tasks? From what I see,  for instance with the `IdleStateHandler`, NIO doesn’t hesitate to schedule tasks frequently, so I guess the performance impact isn’t too high?
- Some pool designs enable specifying a delay after which a connection request is failed if the pool didn’t provide one in time, is this something we would also want for a pool?

Thank you!

**Comment by weissi** [comment]
Adding @Lukasa & @tomerd

**Comment by Lukasa** [comment]
Before I dive in further, I just want to say that I think this is a good starting point for a discussion about designs, and I look forward to seeing a bunch of input here.

Some responses:

> I wonder wether the principle of a user provided connection manager makes sense for the Swift NIO HTTP client?

Someday, yes. But I'd begin by not doing that, as it makes it much easier to make changes in the face of testing. It's rarely a good idea to define an interface before you have a specific requirement for a second implementation, as the overlap between the two implementations helps guide design thinking for what the interface should look like.

So I'd begin by implementing just the default pool, and wait for concrete alternative use cases before worrying about how to support those.

> How to give back a connection to the pool?

Start with the simplest design, which is that the code that is done with the connection calls a function on the pool.

> How should we implement the thread safety of the pool?

The pool should not be on an event loop. A pool has nothing to do with I/O, and so shouldn't block the I/O of any loop it's on.

In the future, we may want to try to maintain the locality of connections on an event loop, but an MVP design has the pool guarded by a lock. It may run on a background thread if we're worried about the cost of the pool operations, but even then I'd begin by simply using a lock to get the code to work right before worrying about how it performs.

> Speaking of the data structure storing the connections, could this be as simple as a dictionary that keys the connections per host/scheme?

So long as you're careful to normalise the scheme, yes. Again, in future this data structure will want to evolve in complexity, but beginning with something stupid is usually a good plan.

> How does this « global cleanup interval » strategy compare to scheduling timeouts per connection using NIO tasks?

If the pool inserts `IdleStateHandler`s in all connections then by default this system will basically "just work", so I'd start there. Again, no need to put the cart before the horse: do the thing that is stupidest and easiest to ensure we get right, and then worry later about whether it performs well.

> Some pool designs enable specifying a delay after which a connection request is failed if the pool didn’t provide one in time, is this something we would also want for a pool?

For v1, no. Maybe later.

**Comment by weissi** [comment]
Thanks @adtrevor for starting to think about this. This is a key thing for what the community needs next!


> So I'd begin by implementing just the default pool, and wait for concrete alternative use cases before worrying about how to support those.

+1, I think starting with something minimal that we land and then iterate on is the winning strategy. As Cory points out, we can define an interface later when we learned much more about connection pooling and what the community really needs.

> > How should we implement the thread safety of the pool?
> 
> The pool should not be on an event loop. A pool has nothing to do with I/O, and so shouldn't block the I/O of any loop it's on.
> 
> In the future, we may want to try to maintain the locality of connections on an event loop, but an MVP design has the pool guarded by a lock. It may run on a background thread if we're worried about the cost of the pool operations, but even then I'd begin by simply using a lock to get the code to work right before worrying about how it performs.

100% happy with what Cory proposes but we could also start with an EventLoop-local connection pool. Ie. one that pools connections _per EventLoop_. So if you request a connection from an EventLoop that doesn't have an existing connection in the pool, it would create a fresh one. Then you wouldn't need any locks but obviously we wouldn't keep the connections as hot as possible :).
I'm really happy either way.

tl;dr: Start with something really dumb first, we can always evolve. I would bootstrap it in this package, at the beginning it can remain totally private so we can change implementation and API at any time.

**Comment by tanner0101** [comment]
> we could also start with an EventLoop-local connection pool. Ie. one that pools connections per EventLoop. 

This is how [Vapor 4's ConnectionPool](https://github.com/vapor/nio-kit/blob/master/Sources/NIOKit/ConnectionPool.swift) operates. A nice side effect as mentioned is that we don't need to worry about locking and the implementation is quite simple.

From Vapor's perspective at least, connection pools should always be associated to a single event loop. An HTTP request coming into a handler may need to pull connections from multiple pools--maybe Redis, PostgreSQL, and HTTP client for example. If each of these pools has connections associated with different event loops, that could incur unnecessary loop hopping. 

However, for the sake of HTTPClient, I don't think the same restrictions apply. Vapor for example always [initializes HTTPClient](https://github.com/vapor/vapor/blob/master/Sources/Vapor/Services/Services%2BDefault.swift#L33) using `EventLoopGroupProvider.shared` with a single event loop anyway. So regardless of how HTTPClient implements its connection pool, Vapor won't have to worry about extraneous loop hopping.

For people using HTTPClient directly with a supplied ELG, a pool that spans multiple event loops could actually be more powerful. For example, if you supply HTTPClient an ELG w/ 8 loops, and you want to configure HTTPClient to have at most 4 connections open at once, you could not have one pool per event loop. You would need a pool that could span the event loops. I'm not sure this is something we need to support, but it's worth considering. (This is a limitation in Vapor fwiw).

**Comment by adtrevor** [comment]
@weissi @Lukasa @tanner0101 Thank you for your feedback! 🙂 

@Lukasa  Nice to know many things can and should be kept as much simple for starting!

> > Speaking of the data structure storing the connections, could this be as simple as a dictionary that keys the connections per host/scheme?
>
> So long as you're careful to normalise the scheme, yes.

Do you refer to case normalisation etc... ?

> If the pool inserts IdleStateHandlers in all connections then by default this system will basically "just work", so I'd start there.

That's great! Hadn't realised it could be enough!


@weissi If we used an EventLoop local pool, it would mean that we may have multiple pools per client, right? If so, does this mean that it would be harder to ensure that a maximum concurrent connections per host limit is respected? Maybe I'm missing something, so here is an example: if a request to a server is given to event loop 1, but a subsequent request to the same server is given to event loop 2, do we now have two separate pools that both hold connections to the same server?

If I understand correctly the source code you posted @tanner0101,  a `ConnectionPoolSource` is not only associated with a specific event loop but also generates connections to a specific host and not multiple ones, right? Do you have a link to an example of how it is used in practice?

**Comment by tanner0101** [comment]
> a ConnectionPoolSource is not only associated with a specific event loop but also generates connections to a specific host and not multiple ones, right? 

@adtrevor yes exactly. Each connection in the pool must be the same (connected to the same host, port, same auth state, etc)

For HTTP client, I imagine you'd have a dictionary where host/port is the key and value is a pool. Here are some examples of how the connection pool API looks like in use:
https://github.com/vapor/nio-kit/blob/master/Tests/NIOKitTests/ConnectionPoolTests.swift

**Comment by Lukasa** [comment]
> From Vapor's perspective at least, connection pools should always be associated to a single event loop. An HTTP request coming into a handler may need to pull connections from multiple pools--maybe Redis, PostgreSQL, and HTTP client for example. If each of these pools has connections associated with different event loops, that could incur unnecessary loop hopping.

The downside of this approach is that it performs poorly in some workload types.

In particular, in cases where a connection is needed with moderate frequency (e.g. every 2s) but for short times (e.g. a few hundred ms), it is highly possible that systems with large numbers of event loops will spend unnecessary resources creating and maintaining many times more connections than are actually required. For example, if you have 16 event loop threads, you will need to create 16 connections to support this case, despite the fact that these connections spend the vast majority of their time idle (average connection idle time is 32s, as only one connection is needed every 2s). Worse, these connections may be idle long enough to be *timed out*, which leads to the connection pools having an effective size of zero and providing no pooling at all.

In general the "pool per event loop" model has troubling properties as connection utilisation drops or as loop count increases. It's a useful model for *some* use cases, but not for all. This is why, down the road, we will probably need to invest in multiple pooling models. However, for now, we are best served building only one.

I don't much mind whether we build the loop-local or global pool first. The advantage in my mind of building the global pool first is that it requires all consumers of the code to ensure that they are guaranteeing their own thread-safety, rather than allowing code to be written that accidentally relies on that fact and cannot be easily tightened up when using a global pool. But I don't care strongly, and if we can move faster by building a per-loop pool then that's fine too.

> > > Speaking of the data structure storing the connections, could this be as simple as a dictionary that keys the connections per host/scheme?
> > So long as you're careful to normalise the scheme, yes.
> Do you refer to case normalisation etc... ?

No, I refer to *key* normalisation. This includes case normalisation, but is not limited to it.

> For HTTP client, I imagine you'd have a dictionary where host/port is the key and value is a pool.

We're beginning to overload the phrase "connection pool", however, so we should aim to clarify. When I have used the phrase "connection pool", I am referring to an object with the following member function on it:

```swift
func obtainedPooledConnection(_ key: PoolKey) -> EventLoopFuture<Channel>
```

What @tanner0101 is describing as a "connection pool" is a sub-component of the above that applies within a single key, and it has a function like:

```swift
func obtainConnection() -> EventLoopFuture<Channel>
```

We should endeavour to distinguish these two things. I'd make the case that we should call the outer layer the "connection pool", as it's the correct level of abstraction for the rest of swift-nio-http-client to operate on. The internals of that object should not be considered relevant for the rest of the client.

Also, re "For HTTP client, I imagine you'd have a dictionary where host/port is the key": that's insufficient for a HTTP client. In general on this issue we should aim to cleave as closely as possible to the [fetch](https://fetch.spec.whatwg.org/#connections) specification for web browsers, as that behaviour will reduce the risk of users seeing surprising or insecure behaviour. We need to take it with a grain of salt as many things will be unnecessary outside of a browser context, but it's a good start.

Fetch wants the pool key to be a tuple of *origin* and *credentials* where *credentials* indicates whether the connection is associated with whether the requests on that connection have been credentialed (e.g. with cookies or Auth headers). Right now swift-nio-http-client is not well placed to make this determination so we can potentially omit the credentials aspect to get off the ground, but we should revisit it.

However, note that the other part of the key is not host/port, it's *origin*. Origin is a 4-tuple of (scheme, host, port, domain). We can omit domain from our notion of origin because it's largely not relevant to us (it'll always be present and always be equal to host), so we want the three-tuple of (scheme, host, port). It's vital to include scheme here as `http` and `https` connections must never be treated as the same. As we aren't a browser, we don't need the other set of schemes they support in our pool, so for us `scheme` is a two-case enumeration.

**Comment by adtrevor** [comment]
@tanner0101 Thanks! Interesting to see how you implemented it for Vapor, I also like the `pool.withConnection { ... }` method.

@Lukasa While it's my first time implementing a connection pool, the global pool is something that I find easier to reason about in terms of performance benefits compared to having multiple pools. FWIW I also believe that my personal use case for this lib (web crawling) would likely encounter the performance issues you are mentioning if using event loop local pools.

> The advantage in my mind of building the global pool first is that it requires all consumers of the code to ensure that they are guaranteeing their own thread-safety

What does "all consumers of the code" refer to? Doesn't the pool handle itself it's thread safety using locks or a specific thread? Were you just thinking about `Future` hopping?

> We're beginning to overload the phrase "connection pool" [...]

Thanks for the clarification. 
I see you gave the example of a `func obtainedPooledConnection(_ key: PoolKey)` method to which we pass a `PoolKey`. Does this mean we want the pool to expose the keys to the outside and require callers to init the keys themselves? The alternative I had thought of was maybe to build it internally and instead let callers get a connection to a server by just passing a URL? What would be best?

Also, while a connection is a `Channel`, do we want to directly return a `Future<Channel>` from the pool, or to instead return a channel warper object that would for instance contain the key the channel is associated with? The issue (?) with directly returning a `Channel` is that when the pool consumer releases it, we cannot reconstruct a `PoolKey` from a channel alone AFAIK, so it may require hacky solutions for it to work properly.

Regarding the connection providing component (the one on which we'd call `func obtainConnection() -> EventLoopFuture<Channel>`), would we essentially have two kinds of it, one for HTTP/1.1 providing concurrent TCP connections to the same server, and another for HTTP/2 providing multiplexed connections?

Thank you

**Comment by Lukasa** [comment]
> What does "all consumers of the code" refer to? Doesn't the pool handle itself it's thread safety using locks or a specific thread? 

While the pool does, the users of the `Channel` may not. If users know that the `Channel` they receive is always on the same event loop to them, they may be incautious about the thread-safety of communicating with that `Channel`. For example, they may communicate directly between `ChannelHandler`s, or avoid using `EventLoopFuture.hop(to:)`. That's not ideal: in some senses it's better to have the more adversarial environment be the first one, at least from a correctness perspective.

> The alternative I had thought of was maybe to build it internally and instead let callers get a connection to a server by just passing a `URL`?

That's certainly fine: I was only drawing out the fundamental portion of the API. It's not a problem at all to provide an even-higher-level one instead.

> Also, while a connection is a `Channel`, do we want to directly return a `Future<Channel>` from the pool, or to instead return a channel warper object that would for instance contain the key the channel is associated with? 

You're quite right, we probably want a `HTTPConnection` object that provides this kind of abstraction.

Regarding the connection providing component (the one on which we'd call `func obtainConnection() -> EventLoopFuture<Channel>`), would we essentially have two kinds of it, one for HTTP/1.1 providing concurrent TCP connections to the same server, and another for HTTP/2 providing multiplexed connections?

At the top level, probably not. The issue is that you don't know which one of those things you'll have to be until you've made your first outbound connection. So you'd probably want an abstract one that is backed by an `enum` that stores the concrete pool type.

**Comment by adtrevor** [comment]
@Lukasa 
Thank you! With all of this, I think I now have a pretty good starting point to prototype an implementation.

> So you'd probably want an abstract one that is backed by an enum that stores the concrete pool type.

At first I thought I'd have a protocol such as 
```swift
protocol ConnectionProvider {
    func obtainConnection() -> EventLoopFuture<Channel>
}
```
And, on first connection request to a server, determine what protocol to use and instantiate the correct concrete type conforming to `ConnectionProvider`.
Were you proposing an `enum` for the potential performance gain?

**Comment by Lukasa** [comment]
> Were you proposing an `enum` for the potential performance gain?

Yes.

Your version will require the use of a protocol existential, which is not ideal when we know the exact concrete types we need.

**Comment by adtrevor** [comment]
My latest PR is still extremely WIP, many things can probably be improved and I'll be happy to get your feedback on this! 🙂

**Comment by adtrevor** [comment]
Additionally, the HTTP/1.1 connection provider doesn't remove itself automatically from the providers list, timeouts are not correctly implemented yet, and again other stuff must still be missing. What I'm most interested by is to know if there is no major design flaw with my approach.

**Comment by crontab** [comment]
Great to see some progress here! Potentially stupid questions:

• Shouldn't HTTP2 fall back to HTTP1 automatically? And therefore the pool could deal only with HTTP2 connection providers? Just for the sake of keeping things simpler.

• Does the ConnectionPool module handle broken channels properly? I.e. it shouldn't return problematic or broken connections back to the pool. I think this is potentially a big problem: some pool users may be struggling with reusing broken/stale/whatever connections where opening a new one would work fine.

I'd be happy to contribute once I get a better grasp of NIO. Just started reading about it today :)

**Comment by Lukasa** [comment]
> Shouldn't HTTP2 fall back to HTTP1 automatically?

What do you mean by this?

**Comment by crontab** [comment]
> > Shouldn't HTTP2 fall back to HTTP1 automatically?
> 
> What do you mean by this?

What if the server doesn't support HTTP/2? Shouldn't the client fall back to 1?

**Comment by adtrevor** [comment]
@crontab 
> Shouldn't HTTP2 fall back to HTTP1 automatically?

Do you mean not using the pool on HTTP/1.1?
IMO if the pooling mechanism is correctly designed this shouldn't be too much of an issue. Additionally much of the web is still HTTP/1.1 so it might not be a good idea skipping pooling for it.

@crontab @Lukasa @weissi  I pushed a new version of my PR that should fix many of the issues @Lukasa pointed out, again that's still a WIP but I believe it should be more readable so don't hesitate to tell me if you have new feedback.

One of the issues I encounter right now is actually related to what @crontab mentioned : I'm not exactly sure of how to manage broken connections.
On HTTP/2 the current approach simply fails right away all all of tasks associated to the child channels (multiplexed streams). This seemed like the simplest approach to implement that would behave in a valid, yet maybe not optimal way, as you are pointing out @crontab.

The HTTP/1.1 provider keeps track of two things: the number of leased connections, and the non-leased ones are kept in a buffer.
- If a non-leased connection closes, it is not automatically restored. Instead, it is simply removed from the buffer.
- If a leased connection is closed when released, the provider won't restore it.

If, at any point in time, there are no more leased connections and the non-leased ones buffer is empty, the provider marks itself as closed and fails any potential tasks that would still be waiting. It seems to me that a good behaviour naturally emerges from this, in particular:
- If no one has needed the connections for a long time, all previously used connections are now in the non-leased buffer, and once they reach their timeout, they close, causing their parent provider to "close" in return and be removed from the pool.
-  If there is a major network failure, all connections will close, there might still be remaining tasks waiting for the connection, and they will be immediately notified of the failure, letting the HTTP Client user do what they think is best (for instance, reschedule a try in the near future).
- In case of partial failure, the tasks that can still be completed are so, and while the number of concurrent connections might drop temporarily, new connection requests will quickly restore a good concurrency.

Back to how the HTTP/2 provider currently works and how it could be improved, I thought about having a trying to restore the connection a few times, maybe with slightly increasing cool down delay between each try. Honestly I don't know at all if that's worth it or not: IMHO, the nice thing with the approach I took for HTTP/1.1 is that the client is quickly notified of failure, and they are responsible of what they prefer to do in this case, so why wouldn't it be the case for HTTP/2?

**Comment by adtrevor** [comment]
@Lukasa Regarding putting HTTP/2 on a separate PR that's what I'll do but I'm currently keeping the code there so that it evolves with the rest, and will remove it just before merging.

**Comment by adtrevor** [comment]
@crontab 
> > > Shouldn't HTTP2 fall back to HTTP1 automatically?
> > 
> > 
> > What do you mean by this?
> 
> What if the server doesn't support HTTP/2? Shouldn't the client fall back to 1?

That's exactly what it will do: the protocol to use is negotiated though [ALPN](https://fr.wikipedia.org/wiki/Application-Layer_Protocol_Negotiation) and the right connection provider is than used.

**Comment by Lukasa** [comment]
> What if the server doesn't support HTTP/2? Shouldn't the client fall back to 1?

Yes it should. However, at some point in the stack the connection pool needs to notice that the connection it’s holding uses HTTP/1 instead of HTTP/2, because the pooling logic is different. 

@adtrevor

Generally the logic sounds ok, but why does having no more open connections fail pending work? Why aren’t we trying to establish more connections?

**Comment by crontab** [comment]
> the nice thing with the approach I took for HTTP/1.1 is that the client is quickly notified of failure, and they are responsible of what they prefer to do in this case, so why wouldn't it be the case for HTTP/2?

I don't think you can transparently recover a HTTP/2 connection with potentially tens or hundreds of streams left in different states each. Or do you mean a failure in just one stream?

Speaking of which, one possibility is reaching the maximum number of streams, where a new HTTP/2 connection should be created by the pool transparently (unless it's what HTTP2StreamMultiplexer does for you anyway?)

**Comment by Lukasa** [comment]
> Speaking of which, one possibility is reaching the maximum number of streams, where a new HTTP/2 connection should be created by the pool transparently 

To begin with, we should disregard this use-case. The stream ID pool is very large: there are 2<sup>30</sup> stream IDs available to a client, which is more than 1 billion. If you've managed to make 1 billion requests over a single TCP connection that's frankly sterling work, and you deserve the right to file the first bug about us not supporting recycling that connection.

**Comment by Lukasa** [comment]
Unless you meant reaching the value of `SETTINGS_MAX_CONCURRENT_STREAMS`, in which case we still should not create a new connection. In fact, RFC 7540 says "SHOULD NOT", which is shouty-case and so very important.

**Comment by crontab** [comment]
> Unless you meant reaching the value of SETTINGS_MAX_CONCURRENT_STREAMS

Yes, it's what I mean. The default for nginx for example seems to be 128 which is not such a fantastical number to achieve for example between microservices physically close to each other.

**Comment by Lukasa** [comment]
Regardless, this is a SHOULD NOT for a reason, which is that `SETTINGS_MAX_CONCURRENT_STREAMS` is intended to be a setting that allows services to exert back pressure on peers attempting to overwhelm what that service can achieve in terms of concurrency. We should respect the signal the peer sends: if they're sending a bad signal, that's their configuration issue and should be addressed there.

**Comment by crontab** [comment]
@Lukasa that's good to know actually because now I realize that for some use cases your backend may not necessarily need a client pool. Typically, backend systems talk to a known set of 3rd party services, you may just need to create one HTTP/2 connection per each in your backend app. Even services that don't directly support HTTP/2 usually hide behind nginx these days. Some recovery logic on your backend may still be necessary though.

But of course for some more general use cases ConnectionPool might be very useful.

**Comment by adtrevor** [comment]
@Lukasa 
> Generally the logic sounds ok, but why does having no more open connections fail pending work? Why aren’t we trying to establish more connections?

If you have a large amount of pending work, but are in a situation where proper connections cannot be established anymore, I was worried that the repeated, potentially large amount of attempts in a short period of time would cause a too high amount of pressure on both sides of the connection. For instance let's say we have a thousand connection attempts in a short period of time, can this have a noticeable performance impact on the client? And if the many requests reach the server (but it cannot accept them), could this be considered as an agressive behaviour?

Obviously, if this truly can cause too much pressure on either side, the connection attempt frequency / max count can be set to stay reasonable, but in that case what's the argument for having this kind of complexity built-in, rather than to delegate it to the client user so that they can choose what they want to do themselves?

@crontab Sorry I realise I made a typo in:
> the nice thing with the approach I took for HTTP/1.1 is that the client is quickly notified of failure, and they are responsible of what they prefer to do in this case, so why wouldn't it be the case for HTTP/2?
I meant "is that the client user" instead of "is that the client".

**Comment by Lukasa** [comment]
It’s fine to have back off logic on connection attempts, but we shouldn’t abandon it entirely.

**Comment by adtrevor** [comment]
@Lukasa Do you agree that the right approach here for retries is could be to put this logic outside of the pool itself? I might be wrong, but I don't see any elegant design where we could make this entirely transparent for the consumer, so it probably should be on consumer side?

**Comment by Lukasa** [comment]
Yes. The connection pool does not own retries or the retry logic. In general, resist the urge to attempt to fold anything more than low-level connection management into the pool or it will rapidly grow impossible to actually write the code for it.

Fully fledged HTTP clients are fiendishly complex, not least because they have multiple interlocking pieces. Retries are one of those pieces. They need information *from* the connection layer (e.g. they need to know the kind of errors that happened, particularly if they're connection based), but they don't want to be integrated into the connection layer, because they also need to consider higher-level semantic issues (e.g. 503 with Retry-After header).

Resist the urge to merge things together. The connection pool's job is to service the query of "I need a connection to origin X on scheme Y, can you provide it?". Nothing else.

**Comment by adtrevor** [comment]
@Lukasa Thank you!
Following this "role separation" principle, the approach I have in mind is to have the `ConnectionPool` always try to restore connections that would have closed as long as there are consumers waiting for one. Another component will be responsible for both re-trying requests and applying a back off logic to keep a sane retry count / frequency. Passed a certain limit, this component would consider this `(host, scheme, port)` as having an issue, and thus would propagate the right error to all associated tasks, closing the associated `ConnectionProvider` at the same time (from this point, it will be up to the HTTP Client _user_ to decide how to handle the error).

Finally, the current implementation causes the buffered connection requests (the ones that can only be treated later in order to respect concurrency rules) to be treated in FIFO order, concretely this means that a failed request and its corresponding retry may not be sent consecutively, but be separated by many other unrelated requests to this (host, scheme, port). This seemed a little counter intuitive at first, but I see no concrete reason why this would be problematic but would of course like to know if you believe that's a correct assumption 🙁

**Comment by Lukasa** [comment]
> This seemed a little counter intuitive at first, but I see no concrete reason why this would be problematic but would of course like to know if you believe that's a correct assumption 🙂

Yup, this is the correct behaviour. In general what you want to do is to enable support for "cancellation": for allowing the retry manager to tell the connection pool that it is no longer interested in the result of the connection request. If the connection pool has no other outstanding requests for that three-tuple it can abandon the connection attempt entirely: otherwise, it can simply serve the next request.

**Comment by adtrevor** [comment]
Thanks!
About request retries caused by a semantic errors, the goal must be to keep this transparent from the point of view of the `HTTPClientResponseDelegate`? For instance if the retry manager catches a 503 error, and the retry is a success, it's only at this point that the delegate is notified of response content?
The issue if we do this is that the delegate is also notified of a _sent_ request through `func didTransmitRequestBody(task: HTTPClient.Task<Response>)`, but here, we can hardly be transparent: if the request fails on a semantic level, the delegate will have already been notified that it has been sent.

I don't know what to do in this case?
- Sending the _request body sent event_ only when we are sure there is no semantic issue obviously requires having already received the body, which seems undesirable?
- Sending the event multiple times isn't a good option either AFAICT: it necessarily makes the retry policy non-transparent to the user.

**Comment by adtrevor** [comment]
Just realised that `didTransmitRequestBody(task: HTTPClient.Task<Response>)`:
- Is called `write()` is called on the `TaskHandler` without waiting on the write promise future: looks like a bug?
- Will be called multiple times in case of a redirection

So if that's intentional we can probably do the same for retrying the request, but I'm pretty sure it's not?

**Comment by Lukasa** [comment]
> About request retries caused by a semantic errors, the goal must be to keep this transparent from the point of view of the `HTTPClientResponseDelegate`? For instance if the retry manager catches a 503 error, and the retry is a success, it's only at this point that the delegate is notified of response content?

I think ideally it would be transparent, yes.

> The issue if we do this is that the delegate is also notified of a sent request through `func didTransmitRequestBody(task: HTTPClient.Task<Response>)`, but here, we can hardly be transparent: if the request fails on a semantic level, the delegate will have already been notified that it has been sent.

This callback appears to be intended to allow the user to drop any resources they might have held in order to generate the request body. For that reason it would be acceptable to delay that notification until we have a non-retryable state, but we should probably clarify the goal of that callback. Given your above note about being called multiple times anyway it would probably be good if @artemredkin could clarify the intended purpose of this callback.

> Sending the request body sent event only when we are sure there is no semantic issue obviously requires having already received the body, which seems undesirable?

Not necessarily, doing it when you've got the response head should be sufficient.

**Comment by adtrevor** [comment]
@Lukasa Thank you!

I would indeed be curious to know what you think of this, @artemredkin? In fact, the problem also applies to other delegates methods in the case of re-trying idempotent requests : `didReceivePart` would be called again on re-try for instance. This is a case where the retry mechanism cannot be made entirely transparent.

Maybe it would be a good thing to add an additional delegate method for indicating that an event causing a new request to be sent? This could play nicely with new redirect options allowing for a redirection to take place only after the body has been transmitted and not immediately after head has been parsed, as it currently is.

**Comment by adtrevor** [comment]
Why not something like this?
- Both redirections and retries are entirely transparent when using the `execute(request: Request, timeout: Timeout? = nil) -> EventLoopFuture<Response>` method (that uses the `ResponseAccumulator` internally)
- When a user provided delegate is used, we could avoid delaying anything, and instead just rely on the new delegate method to allow detection of retries and redirections.
- Additional options could be about managing wether or not the body should be downloaded (in the case of some 3xx/5xx codes, for instance) as well as what HTTP codes the user wants to be automatically handled by the client.

**Comment by artemredkin** [comment]
`didTransmitRequestBody` is needed to notify the client that request was sent. It is indeed not ideal that it's called before write succeeds. In the future this callback will be changed a bit and will be called, similar to `didReceivePart`, on every sent body chunk (so one can display upload progress bar, for example).

**Comment by artemredkin** [comment]
I'll need to think about what to do in case of retries...

**Comment by artemredkin** [comment]
Maybe we can introduce a new method -> `requestRetried`, that will give designer of the delegate a way to reset internal state. wdyt?

**Comment by adtrevor** [comment]
@artemredkin I like it but maybe it would also be useful for the delegate to be notified of a redirection? In this case do you think it would be better to have a single method covering everything that causes a new request to be sent and would have an argument indicating the reason and additional details?
Or is it better to have separate methods? 

Thank you

**Comment by artemredkin** [comment]
Yes, having a notification about redirection is also a good idea. I haven't yet gotten to improvements to redirection yet, so I don't have any concrete ideas. One thing to consider is that in case of redirection we might want to decide whether to redirect or not, unlike retry (but maybe in case of retry too). So having two methods, one for retry and one for redirect seem like a better way, but I'm not super sure about that.

**Comment by artemredkin** [comment]
@adtrevor btw, thanks for pointing the issue with `didTransmitRequestBody`, filed https://github.com/swift-server/swift-nio-http-client/issues/38 to track!

**Comment by adtrevor** [comment]
@artemredkin, cc @Lukasa 
In #21 you had given this example:
```swift
protocol HTTPDelegate {
    func willRedirect(to url: URL) -> Bool
}
```

This makes me think that maybe it would be a good idea to split method requirements across multiple delegates where this makes sense? In such a design we would keep `HTTPClientResponseDelegate` for informative events, and add new delegates such as `HTTPRedirectDelegate` and `HTTPRetryDelegate` to influence the HTTP Client behaviour. This would allow mixing user implemented parts with client implemented parts (for instance, the user provides an `HTTPDelegate` for custom redirects, and the client uses its own `HTTPClientResponseDelegate`, such as `ResponseAccumulator`). What do you think about this?

If this approach makes sense, `HTTPClientResponseDelegate` would mainly be focused on resetting its internal state when receiving a "new request sent" event.

I'm not really sure either of whether it's better or not to have all events sent to the same method or to multiple ones.
We should note that some `HTTPClientResponseDelegate` implementations don't need to distinguish the actual reason why a new request was sent, that's the case of `ResponseAccumulator` AFAICT.

Another point to take into account is how future changes, and in particular adding reasons for sending new requests, might affect both approaches:
- If we add multiple methods for each kind of "new request sent" event, we would probably provide empty default implementations for them, as it's currently the case for most of the `HTTPClientResponseDelegate` methods. In this case, adding a new reason *isn't* a source breaking change.
- If we add a single method, we'll probably inform the delegate through an enum passed as an argument to this method. In this case, adding a new reason *is* a source breaking change.

If we go with the first option, some users might do the following when implementing their delegates:
```swift
func didRedirect() {
    resetState()
}

func didRetry() {
    resetState()
}

func resetState() {
    // Same code for both events
}
```
If the library ever adds a new method (with a default implementation), it won't be a source breaking change in the sense that everything will still compile fine, but it will definitely break the behaviour of the delegate as the newly added method will likely need be taken into account (for instance, giving it an implementation that calls `resetState()`).

This kind of issue is much less likely to occur on a single method: the same method will also be called for future new reasons, and if the implementation had its behaviour influenced by the reason and switched over it, a _non exhaustive switch_ error will immediately occur when the user starts using the new library version, letting them quickly take the new cases into account.

In the end, we could also simply avoid providing default implementations if we use multiple methods so that newly added methods are guaranteed to be noticed by the user. It would mean users couldn't ignore these methods altogether anymore when they don't need them? Maybe that's not a real issue and all sensible implementations should always take them into account anyway?

One last option I see that could also solve it is for future "automatic new request reasons" to be opt-in, meaning that the user would have to manually enable them in the client option, and therefore be aware of what behaviour change they would get. But I'm not really sure it's a great option either.

**Comment by Lukasa** [comment]
In general I'm in favour of splitting the delegates up: the only problem ends up being that you store multiple references to them, but that's not really the end of the world.

**Comment by artemredkin** [comment]
Sounds reasonable. @Lukasa where do you think this be an issue? In `HTTPTaskHandler` where we will have references to all delegates? Or do you think this will be an issue for the clients using this API? The latter could be address by having one class implementing all delegates, no?

**Comment by Lukasa** [comment]
Regardless of how the user handles it, the downside of having multiple delegates is only that you hold multiple references, either to separate objects or the same object. As I stated, it mostly doesn't matter. :wink:

**Comment by adtrevor** [comment]
Great! Thanks

**Comment by weissi** [comment]
closed with #105
